runtime/lua/vim/_editor.lua

```diff
@@ -789,7 +790,7 @@ function vim.deprecate(name, alternative, version, plugin)
   plugin = plugin or 'Nvim'
   message = alternative and (message .. ', use ' .. alternative .. ' instead.') or message
   message = message .. ' See :h deprecated\nThis function will be removed in ' .. plugin .. ' version ' .. version
-  vim.notify_once(message, vim.log.levels.WARN)
+  vim.notify_once(debug.traceback(message, 2), vim.log.levels.WARN, nil, message)
```

Comment:
i think we need a `fatal` param. 

1. `fatal=false` : "soft" deprecation, emits at WARN level, no backtrace, and don't fail (the implementation still works)
2. `fatal=true` : "hard" deprecation, calls `error()` rather than `vim.notify_once`. Don't need debug.traceback.

With a "hard" deprecation there's no point in only showing the error once because the implementation doesn't exist ...

Reply:
But the entire point of `vim.deprecate()` was to *not* abort the calling code with errors. If it's a fatal error then you could just use `error()` instead. The code that is calling a deprecated function should still work before it's actually removed. It's just that you should see exactly where the deprecated function is used.

Maybe it's okay to reuse `vim.depracate()` to throw actual errors too, but I think the backtrace is needed for a "soft" deprecation too.

Reply:
> But the entire point of `vim.deprecate()` was to _not_ abort the calling code with errors.

No. The minimum use-case of vim.deprecate() is to have a common shape for deprecations (i.e. the message has the same wording, mentions alternative, optionally mentions version...)



> If it's a fatal error then you could just use `error()` instead.

Some small downsides of error() are 

1.  message isn't in the "standard" format
2. loses the theoretical opportunity to statically analyze `deprecate()` calls
3. loses theoretical opportunity to localize the message

> The code that is calling a deprecated function should still work before it's actually removed.

I plan to use vim.deprecate() for hard removals. It's just easier than hand-craft the usual message.

Reply:
> Maybe it's okay to reuse vim.depracate() to throw actual errors too, but I think the backtrace is needed for a "soft" deprecation too.

We can try it. It's a reversible decision :)

Reply:
> Show a backtrace only if `vim.o.verbose > 0` ? 

There can be some way of disabling it, but I think showing a backtrace should be the default. I guess verbose level cannot be a negative number?

> The idea is that some users will want to continue to work and can't update stuff, so showing a backtrace all the time will be annoying.

The message is only shown once per neovim instance, so it's not all the time. The argument could be made that backtraces will force you into the `Press ENTER or type a command to continue` prompt, but `vim.deprecate` is already a multiline message so it does that either way.